sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,513 @@
+`range` iteration space of `Dims` dimensions.
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member variables and type
+aliases:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+static constexpr int dimensions = Dims
+----
+!
+The number of dimensions of the kernel's range.
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+
+a|
+*`nd_range_kernel` property*
+
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+struct nd_range_kernel_key {
+  template <int Dims, typename Name>
+  using value_t = property_value<nd_range_kernel_key, Dims, Name>;
+};
+
+template<int Dims, typename Name = void>
+inline constexpr nd_range_kernel_key::value_t<Dims, Name> nd_range_kernel;
+
+template<>
+struct is_property_key<nd_range_kernel_key> : std::true_type {};
+----
+!====
+
+Indicates that the function is a raw kernel that is invoked with an
+`nd_range` iteration space of `Dims` dimensions.
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member variables and type
+aliases:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+static constexpr int dimensions = Dims
+----
+!
+The number of dimensions of the kernel's range.
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+
+a|
+*`single_task_kernel` property*
+
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+struct single_task_kernel_key {
+  template <typename Name>
+  using value_t = property_value<single_task_kernel_key, Name>;
+};
+
+template<typename Name = void>
+inline constexpr single_task_kernel_key::value_t<Name> single_task_kernel;
+
+template<>
+struct is_property_key<single_task_kernel_key> : std::true_type {};
+----
+!====
+
+Indicates that the function is a raw kernel that is invoked via `single_task`
+(i.e. without any iteration space).
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member type alias:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+|====
+
+When a function is defined as a raw kernel, each parameter to the function is
+a kernel argument, which must abide by the rules for allowable kernel parameter
+types specified in section 4.12.4 "Rules for parameter passing to kernels" of
+the core SYCL specification.
+The function's return type must be `void`.
+
+The following example demonstrates how a raw kernel can be defined using the
+`range_kernel` property:
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct my_iota;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota>)
+void iota(float start, float *ptr) {
```

Comment:
I really don't like the fact that the user has to specify `iota` and `my_iota` here, and it makes things look really complicated.  I think most users will be confused by this: from their perspective, the function already has a name ("iota"). Could we rewrite the examples so that they don't use a type name?

Rather than make this extension dependent on the online compiler, you could redefine the new kernel bundle functions here.

Reply:
Several of your comments are related to this issue about using a type-name to identify a raw kernel function.  I'll use this place to respond to them all.

My intention was that the syntax like this:

```
kb.ext_oneapi_get_kernel("foo")
```

would only work if the raw kernel was compiled from a string using the [sycl_ext_oneapi_kernel_compiler](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler.asciidoc) extension.

I agree that it would be nice to enable this even if the raw kernel was directly embedded within the single-source SYCL application, however, I'm a little worried it would be hard to implement.  I think it would make sense to work with the development team here.  If there is an easy solution, we can include this feature.

If we do this, though, the user would need to declare the raw kernel as `extern "C"` in order to disable name mangling.  I assume it is not reasonable to require all raw kernels to be defined this way.  What should we do in the cases where the user does not want to declare the kernel function as `extern "C"`?

One option is to invent a new property like `kernel_name<NAME>` where `NAME` is a string.  C++17 does not allow a string to be a template parameter, but I think this is allowed in newer versions of C++.  Even if we allowed this, though, why is this better than using a type-name?  It seems to me that a type-name is probably more convenient than a string here because it's easier to construct a unique type-name for a template function than it is to construct a unique string-name.

Therefore, I think we need some way to let the user assign a type-name to a raw kernel function.  I considered creating a separate property for this, but it seemed too verbose because it would require the user to decorate a function with two properties in order to make it a raw kernel.  This was my motivation for adding the type-name as an extra parameter to the `range_kernel` (etc.) properties.  If we think it will be more common for users to declare raw kernels as `extern "C"`, then maybe this is a good argument for splitting type-name out into a separate property.

Even if we do leave the type-name as an argument to the `range_kernel` (etc.) properties, we could still invent a new property like `kernel_name` that we can use with the new kernel invocation functions.

Reply:
I have a new proposal which I like better than both the current state of this PR and also better than using the string-name to identify the kernel.  Skip to the end of this comment if you want to see the new proposal, but I'm going to start by pointing out problems with the current ones.

I think it's important for the raw kernel extension to support kernel functions even when they are templated or overloaded.  We've seen many people define kernel lambdas or named kernel objects as templates, and it seems reasonable to support overloads also.  The NVRTC feature allows kernels that are both templates and overloads, so it would be good to have that same functionality with the kernel compiler.  Just today, I pushed 0e6f0b226702b65405e8fb41fad588bc811484a8 to explain how the current PR would support templates.  It also supports overloads, but I think the spec does not need any additional explanation for this case because it's pretty straightforward.

I don't really like the current PR's way of handling templates because it requires the user to explicitly instantiate the template for each set of arguments they will use when submitting the kernel.  This seems like an extra step that is ugly and easy to forget.

Your proposed syntax for identifying the kernel by its string-name seems nice on the surface, but I think it breaks down when the argument types (or template types) are not fundamental types.  For example, consider a simple case where the kernel function takes a `size_t` argument:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1>)
void mykernel(size_t sz, float *buf) {
}
```

Because `size_t` is a type alias, the overload is really `void mykernel(unsigned long long, float *)`.  The user would need to know this when calling `get_kernel`:

```
sycl::kernel k = exe_bndl.get_kernel("mykernel(unsigned long long, float*)");
```

You might suggest that the SYCL RT should intrinsically know the definition of `size_t`, but it certainly can't know the definition of user-defined type aliases.  There are similar problems with other user-defined types like enumerations.  Namespaces also add a wrinkle, for example, if the type is defined in some non-root namespace.

This proposal also does not solve the problem with function templates.  The user would still need to explicitly instantiate the template as I describe in 0e6f0b226702b65405e8fb41fad588bc811484a8.

**New Proposal**

My new proposal is to identify the kernel function by its address.  For cases when the kernel function is not templated and not overloaded, the syntax is very simple:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1>)
void iota(float start, float *ptr) {
   // ...
}

void foo() {
  sycl::kernel k = exe_bndl.get_kernel(iota);
}
```

Note that this works even without defining the function as `extern "C"`.

For more complicated cases, you can use normal C++ syntax to get the function pointer:

```
sycl::kernel k1 = exe_bndl.get_kernel(my::namespace::iota);            // namespace
sycl::kernel k2 = exe_bndl.get_kernel(iota<int>);                      // template
sycl::kernel k3 = exe_bndl.get_kernel((void (*)(float, float*))iota);  // overload disambiguation
```

There are two other advantages of this approach:

* There is no longer a need to explicitly instantiate the template because the syntax for obtaining the address of the function implicitly instantiates it.
* A raw kernel function can be declared with internal linkage (e.g. `static`) and it all still works because the function address is unique even if there is another function in another translation unit with the same name.

I think we can simplify things even further by avoiding the need to get a `kernel` object.  Instead, the new kernel enqueue functions in [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc) can directly take a raw kernel function pointer:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1>)
void iota(float start, float *ptr) {
   // ...
}

int main() {
  sycl::queue q;
  sycl::parallel_for(q, {N}, iota, 1.f, ptr);  // Note this takes the function pointer and not a `kernel`
}
```

Reply:
This seems much nicer! But it doesn't work if the function pointer isn't a constant expression, correct? If we pass the function pointer as (non-type) template argument, it enforces that it is a constant expression.

Reply:
I agree this seems much nicer, especially with the overload that doesn't require any usage of kernel bundles.

I share Roland's concern about constant expressions, but I wonder if this is something we could just require (via a precondition), rather than making it a template argument.  Passing the function pointer as an argument is much simpler syntax, and we could teach the front-end to complain if the function pointer isn't a constant expression.

Reply:
Actually, I was thinking that the function pointers could be runtime values, and this could be made to work.  However, it would require a runtime lookup on each call to `get_kernel` and `sycl::parallel_for`.  That seems bad, so I agree it would be better to force them to be constant expressions.  The `get_kernel` function would then look like this:

(BTW, I'm going to start using the proposed syntax with `SYCL_EXT_ONEAPI_RANGE_KERNEL` instead of the `range_kernel` property.  See the first issue at the bottom of the document.)

```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) {
   // ...
}

void foo() {
  sycl::kernel k = exe_bndl.get_kernel<iota>();  // Note function pointer is a template parameter now
}
```

We could still have the simplified version of `parallel_for`, but again the function pointer would have to be a template parameter:

```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) {
   // ...
}

int main() {
  sycl::queue q;
  sycl::parallel_for<iota>(q, {N}, 1.f, ptr);
}
```

I agree that the syntax is not as nice, but I still think it's better than requiring the user to get a `kernel` object.

> and we could teach the front-end to complain if the function pointer isn't a constant expression.

We're getting a lot of push-back from the front-end team about adding special error checks, and I think this type of thing is going to make it harder to get SYCL support upstreamed.

I think @rolandschulz is proposing that we do not need any special decoration on the declaration of a raw kernel function.  (I.e. we don't need the `SYCL_EXT_ONEAPI_RANGE_KERNEL` macro or the `range_kernel` property.)  Instead, he says that we can identify that a function is a raw kernel simply by whether it is passed as a template parameter to `get_kernel` or to `parallel_for`.

I agree this is possible, but I'm hesitant for several reasons:

* This would not work when defining the kernel as a string in the [sycl_ext_oneapi_kernel_compiler](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc) extension because the string won't contain any calls to `get_kernel` or `parallel_for`.  I still think this is one of the main use cases for raw kernels, so it seems unappealing to have special syntax that doesn't work in this use case.

* This would place more burden on the front-end to do special SYCL processing because it would need to track the template parameters to `get_kernel`, etc.  Again, I think we want to avoid adding special SYCL knowledge to the front-end when possible.  If we ask users to decorate raw kernels with `SYCL_EXT_ONEAPI_RANGE_KERNEL`, then the macro can expand to a C++ attribute that identifies the function as a kernel.  This is already how the front-end identifies kernel functions today (via the `[[clang::sycl_kernel]]` attribute).

* I think raw kernels are an advanced feature, so I'm not so concerned that people need to add a macro before their declaration.

@rolandschulz is also proposing that the raw kernel function can take an initial argument of `item` or `nd_item`, rather than have the user call the free function to get this.  I'm not convinced this is better.  To me, it seems nice that the number of parameters in the `Args...` pack to `parallel_for` matches the number of function arguments.  I also think it would be less extensible.  Let's say that we wanted to add another kernel parameter like `kernel_handler` in the future.  I think it would be hard to do this in a backward compatible way if `kernel_handler` was an initial parameter to the raw kernel function.  The SYCL RT would need some sort of introspection to see if the function was declared with this extra parameter.  There's no problem with the free functions, though.  We just provide a new free function, and the user can call it if they want to get the `kernel_handler`.

Reply:
Another possible syntax would be: `sycl::parallel_for(q, {N}, kfp<iota>, 1.f, ptr);` (replace kfp with whatever name we decide for this). This has the advantage the argument stays in the same position. Of course it has the disadvantage it requires the wrapping. Not sure which I prefer. 

I'm not convinced by the backward compatibility argument against non-free-function-syntax. This introspection would be trivial to do in the SYCL RT headers. You just take the function arguments and do a constexpr if on the type of the first argument.

Reply:
> I'm not convinced by the backward compatibility argument against non-free-function-syntax. This introspection would be trivial to do in the SYCL RT headers. You just take the function arguments and do a constexpr if on the type of the first argument.

This wouldn't work for the `parallel_for` overload that takes a `kernel` object.

The `kfp<iota>` syntax is interesting.  I also don't know which I prefer.  If we are settling down on the other issues, I will update this spec tomorrow, but leave out the syntax with the new `sycl::parallel_for` (etc.) enqueue functions.  I can add that later after we come to agreement.

Reply:
what do you mean? if it's a kernel object we don't need to pass the kernel arguments because they are members.

Reply:
are you talking about the kfp syntax? if kfp is a variable (not type) you can pass it to this overload.

Reply:
No, this is different from the kfp syntax.  When compiling from a string, you need to use the kernel bundle API to get a `kernel` object that represents the dynamically compiled kernel.  See [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc#example) for an example of how it works when the kernel is written in OpenCL C.  We plan to support the same online compilation technique for kernels written in this "free function kernel" form.

A user can also get a `kernel` object without dynamically compiling from a string by using the kernel bundle API.  I think we should also allow invoking such a `kernel` object using the same overload of `parallel_for` that I list above.

Reply:
> * A raw kernel function can be declared with internal linkage (e.g. `static`) and it all still works because the function address is unique even if there is another function in another translation unit with the same name.

I don't think this will work for offloaded code. The difference is that host application identifies GPU functions by name, not by address.
Let's assume we have two translation units defining static functions with the same name. DPC++ today emits kernel unique name at compile step, but I think unique names for static functions are emitted at the link step.

Greg's comment seems to be a "nice side effect (or bonus) of new proposal" rather than the foundational feature. Whether we can or cannot make free function kernels static should not block new proposal.

Reply:
I think you are describing the current integration header technique that identifies offload kernels by their name.  I was thinking that we could use a technique similar to what we do for `specialiation_id`, where the integration is done by address.  I think we are allowed to have two `specialization_id` variables with the same name in different translation units if they are declared static.

In any case, I do agree that this is not a showstopper for the this proposal.

Reply:
I updated the PR to use the new proposed syntax where a kernel is identified by its function address rather than by a type-name in a7bef5ed34e9e71199da439fa5d27015782c191d.  I did not add syntax yet for enqueueing a kernel by its function address in the new `sycl::paralle_for`, etc, functions only because we haven't agreed on the alternative yet.  However, I did capture this as an open issue (397668c47c09a0b0a6cbe82d05d1601e6ded8187).

My preference is to resolve this conversation, discuss any remaining issues in new conversations.  If anyone wants to keep this conversation un-resolved, make a comment here.  If I don't hear from anyone, I'll resolve it next week.